### DISCUSSION TITLE
How does cookie work between backend and React if it is not returned from the backend upon login?

### DISCUSSION THREAD

**Discussion by juyanghsu** [fastapi-users/fastapi-users#1072] [Q&A] [answered]
This is probably a newbie question so I'm sorry if it doesn't belong here. 

I'm trying to understand how Cookies will work between requests after a user is logged in. I'm using Axios to make a request using the given username and password, which operates and shows that I am successfully logged in. But when I make a second request. It returns a 401 unauthorized error. 

After digging around in the library's source code, I found that there are no cookies returned, so I think I shouldn't be expecting a response. The question is if there is no response for a token or cookie, how do I make a second request to the backend?

```
// first call to login
            axios
            .post("http://localhost:8000/auth/login", form_data)
            .then(function (response) {

                console.log(response);
                console.log(response.data);

                //second call testing if it works
                axios
                .get("http://localhost:8000/users/me",)
                .then(function (response) {
                    console.log(response);

                }).catch(function (error) {
                    console.log(error, "error");
                    console.log(error.response.data.detail)
                    }) // returns a 401 error here

            })
            .catch(function (error) {
            console.log(error, "error");
            console.log(error.response.data.detail)
            });
```

transport and strategy in backend:

```
cookie_transport = CookieTransport()
def get_redis_strategy() -> RedisStrategy:
    return RedisStrategy(redis_client, 3600)

auth_backend = AuthenticationBackend(
    name="view_backend",
    transport=cookie_transport,
    get_strategy=get_redis_strategy,
)
```

Most of the code I have in the backend is similar to the example given in the library. Should I change anything so that I get a cookie or is there something else I should be doing with the front end?
Thanks in advance!

**Comment by Xevion** [comment, marked as answer]
I am operating with a similar stack (Vue, but Redis/Cookie) to you and hadn't thought of actually using the cookies to send a request yet, so I ran into the same issue as you did just now.

The main change I made was just now to add `withCredentials` to my Axios requests. This actually embeds the cookies set on your browser into the request.

```js
return this._vm.$http.post('/api/auth/cookie/login', formData)
                .then(response => {
                    this._vm.$http.get('/api/users/me', {withCredentials: true})
                        .then(response => {
                            console.log(response)
                        })
                })
```

You will probably be running into CORS issues right about now, so make sure you read up on how it works, otherwise, you're going to be wasting a bunch of time tearing your hair out.

What's happening here is essentially the login response has a `set-cookie` header which adds a cookie to your browser session. This cookie contains your login token and it is fiercely protected by your browser to prevent XSS attacks (if someone was to get javascript running on a victim's computer, we need to prevent them from sending authorized requests wherever they want).

But for you to get this working, you need to make sure that `Cookie` header is set with the appropriate token. There's a lot of rules and protections, so read up! Good luck.

[XMLHttpRequest.withCredentials](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials)
[Set-Cookie](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie)
[Missing Cookie - Medium.com](https://medium.com/swlh/7-keys-to-the-mystery-of-a-missing-cookie-fdf22b012f09)

**Reply by juyanghsu** [reply]
Got it, looks like if I were to dig into this more it'll have to take a lot of time to figure things out. 

Without starting a war, I'm most likely going to use JWT token to do things instead and store it in sessionStorage since I won't be using any unsafe HTML, It should be fine as long as I keep it that way.

I'll be marking your response as answer since I think it'll be useful for those who want cookies as method. Thanks for your input!

**Comment by juyanghsu** [comment]
Update:

After going down a rabbit hole on how JWT refresh tokens are still going to be insecure in some cases, I have come around to explore this answer a bit more. The only code that I added just now is when I post with login I included `{withCredentials: True} `, which saved the cookie in the browser, in line with what @Xevion has suggested, the code is done like so:

```
// first call to login
            axios
            .post("http://localhost:8000/auth/login", form_data, {withCredentials: true})
            .then(function (response) {


                console.log(response.headers);

                //second call testing if it works
                axios
                .get("http://localhost:8000/users/me", {withCredentials: true})
                .then(function (response) {
                    console.log(response);

                }).catch(function (error) {
                    console.log(error, "error");
                    console.log(error.response.data.detail)
                    })
            })
            .catch(function (error) {
            console.log(error, "error");
            console.log(error.response.data.detail)
            });
```

I have not seen any CORS issue when trying to call the second function and it properly returned the data I requested for (which does not mean I won't run into them in the future).  If anyone is interested, I will keep updating here to show how this progress. Big thanks to  @Xevion again for providing me with the insights!
